### PR TITLE
add reparameterization, combine sample and sample_n

### PR DESCRIPTION
**Pull request by alicanb**
@apaszke This is a combination of #3886 and #4100. It was easier to redo after #4131 than to fix each PR. I am closing the other PRs.
cc @fritzo, @neerajprad

### CODE REVIEW

File: torch/distributions/gamma.py [hunk 1]
```diff
@@ -61,11 +61,21 @@ def __init__(self, alpha, beta):
         self.alpha = alpha
         self.beta = beta
 
-    def sample(self):
-        return _standard_gamma(self.alpha) / self.beta
+    #def sample(self, sample_shape=()):
+    #	if len(sample_shape) == 0:
```
Comment (neerajprad) [inline torch/distributions/gamma.py]: We can remove these commented lines.

File: torch/distributions/distribution.py [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
-        parameters are batched.
+        Generates a sample_shape shaped sample or sample_shape shaped batch of samples if the distribution
+        parameters are batched. Currently only supports len(sample_shape)<2.
+        """
+        z = self.rsample(sample_shape)
+        return z.detach() if hasattr(z, 'detach') else z
+
+    def rsample(self, sample_shape=()):
+        """
+        Generates a sample_shape shaped reparameterized sample or sample_shape shaped batch of reparameterized samples if the distribution
```
Comment (fritzo) [inline torch/distributions/distribution.py]: Linting is failing because this line is too long. Actually it's especially important to keep docstring width under 80 characters so that `help(foo)` prints correctly inside Jupyter notebooks.

**Review by fritzo — APPROVED** [review]
Looks great! Thanks for implementing this.

File: torch/distributions/distribution.py [hunk 3]
```diff
@@ -1,15 +1,25 @@
...
-    def sample(self):
+    def sample(self, sample_shape=()):
         """
-        Generates a single sample or single batch of samples if the distribution
-        parameters are batched.
+        Generates a sample_shape shaped sample or sample_shape shaped batch of samples if the distribution
+        parameters are batched. Currently only supports len(sample_shape)<2.
+        """
+        z = self.rsample(sample_shape)
+        return z.detach() if hasattr(z, 'detach') else z
```
Comment (apaszke) [inline torch/distributions/distribution.py]: why is this `hasattr` needed?

File: torch/distributions/distribution.py [hunk 4]
```diff
@@ -18,7 +28,8 @@ def sample_n(self, n):
         Generates n samples or n batches of samples if the distribution parameters
         are batched.
         """
-        raise NotImplementedError
+        warnings.warn('sample_n will be deprecated. Use .sample((n,)) instead', PendingDeprecationWarning)
```
Comment (apaszke) [inline torch/distributions/distribution.py]: Did you verify that this warning shows up? We've had some issues where if you mark them as `DeprecationWarning`s, then they are filtered out by default.

File: torch/distributions/distribution.py (hunk 3 above)
Comment (fritzo) [reply torch/distributions/distribution.py]: Distributions sample Tensors if parameters are Tensors or Variables if parameters are Variables. So `z if is_tensor(z) else z.detach()` should also work.

**Comment by neerajprad** [comment]
There are some small merge conflicts again due to the `enumerate_discrete` PR I guess. :) And a linting error. Otherwise, we should be good to go on this.

File: torch/distributions/distribution.py (hunk 4 above)
Comment (neerajprad) [reply torch/distributions/distribution.py]: That might depend on the client settings. In Pyro, we filter out DeprecationWarnings (but emit once), since we found some problematic dependencies which were using some deprecated methods of their own dependencies. 

@alicanb - You could test that the warning is thrown in by using the [catch warnings context manager](https://docs.python.org/3/library/warnings.html#testing-warnings)

**Comment by fritzo** [comment]
@alicanb do you mind if I just push the flak8 fix? I'd really like to build on top of this PR.

**Comment by alicanb** [comment]
Yeah @fritzo go ahead

**Comment by fritzo** [comment]
Thanks, done!

**Review by apaszke — COMMENTED** [review]
LGTM, but please let's add tests for warnings from `sample_n`

**Comment by neerajprad** [comment]
Python 3.6 build is timed, though it looks random to me. Can we restart that test?

**Comment by apaszke** [comment]
@fritzo does it complicate things a lot? We don't need to test it anymore. I assume that it's such a simple thing that tests for sample with different shapes are sufficient

**Comment by fritzo** [comment]
@apaszke I'm just hoping to minimize overhead maintaining backwards compatibility (e.g. it's much easier to refactor torch.distributions than pyro.distributions 😉 ). Is it enough to add a single additional test to this PR?
```py
def test_sample_n_deprecated(self):
    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter("always")
        Normal(0, 1).sample_n(1)
        self.assertTrue(w)
```
Then in a follow-up PR we can replace `.sample_n(n)` with `.sample((n,))` in tests?

**Comment by apaszke** [comment]
Yes, that's exactly what I imagined.

**Comment by apaszke** [comment]
Minus setting the warnings filter. It should be triggered by defualt without fiddling with `warnings` (that's the whole point)

File: torch/distributions/distribution.py [hunk 5]
```diff
@@ -1,25 +1,39 @@
...
 
     def sample_n(self, n):
         """
-        Generates n samples or n batches of samples if the distribution parameters
-        are batched.
+        Generates n samples or n batches of samples if the distribution
+        parameters are batched.
         """
-        raise NotImplementedError
+        warnings.warn('sample_n will be deprecated. Use .sample((n,)) instead', PendingDeprecationWarning)
```
Comment (fritzo) [inline torch/distributions/distribution.py]: Should be a `DeprecationWarning` rather than a `PendingDeprecationWarning` (which are silenced by default).

**Comment by apaszke** [comment]
The warnings don't show up in Python 2

**Comment by alicanb** [comment]
Yeah I just looked at the logs... It seems `DeprecationWarning` is not loud by default in python2.7.

**Comment by apaszke** [comment]
I think we've mostly been using `UserWarning`, as that shows up everywhere by default

**Comment by alicanb** [comment]
Ok I have an idea of what's wrong, we call `sample_n` in tests before checking for warning, so warns the first time we call `sample_n`, but not when we're checking for it. Moving the warning test to the top should solve the problem

**Comment by neerajprad** [comment]
@apaszke - This should be good to merge now.
